### ISSUE TITLE
cuSPARSELt Integration

### ISSUE THREAD

**Issue by calclavia** [pytorch/pytorch#62153]
## 🚀 Feature
Support cuSPARSELt for structured sparse matrix multiplications. (Linear layers)

## Motivation
NVIDIA Ampere GPUs can take advantage of 2:4 sparse pattern via the cuSPARSELt library (https://docs.nvidia.com/cuda/cusparselt/index.html). This can enable almost 2x faster forward pass for certain models that rely on structured sparsity.

Detailed paper: https://arxiv.org/pdf/2104.08378.pdf

## Pitch
A simple API that enables linear layers to execute via the 2:4 sparse pattern using the cuSPARSELt library.

## Alternatives
See: https://github.com/NVIDIA/apex/tree/master/apex/contrib/sparsity
This code simply does the sparse masking, but doesn't provide the cuSPARSELt speedup.

Cupy has an initial/basic implementation, but it seems to be low level CUDA operations.
https://github.com/cupy/cupy/pull/5158

cc @ngimel @nikitaved @pearu @cpuhrsch @IvanYashchuk

**Comment by IvanYashchuk** [comment]
Hey, @calclavia! Thank you for your request!
Currently, the priority is to integrate cuSPARSE library into PyTorch https://github.com/pytorch/pytorch/issues/60854. Once this work is finished we will consider using cuSPARSELt library as well.

**Comment by calclavia** [comment]
@IvanKobzarev Looking forward to it!

**Comment by robogast** [comment]
Hi! Any update on this? I have a specific usecase that benefits from this feature :)

**Comment by kiucho** [comment]
@jcaip Hi, thanks for your working on cuSPARSELt integration.
Is there any prototype or any release I can test for 2:4 sparsity pattern now?

**Comment by jcaip** [comment]
@kiucho We've added support via CUTLASS, which you should be able to use now. See these [docs](https://pytorch.org/docs/main/sparse.html#sparse-semi-structured-tensors) for more information. 

The cuSPARSELt integration will be completed by the 2.1 release, or really when https://github.com/pytorch/pytorch/pull/102133 is merged.

**Comment by kiucho** [comment]
@jcaip Thanks for your great work! I can't wait for pytorch 2.1 to be released or for cuSPARSELt to be available. Thanks again.

**Comment by pkulium** [comment]
> @kiucho We've added support via CUTLASS, which you should be able to use now. See these [docs](https://pytorch.org/docs/main/sparse.html#sparse-semi-structured-tensors) for more information.
> 
> The cuSPARSELt integration will be completed by the 2.1 release, or really when #102133 is merged.

Thanks for this work! I am trying to speed up facebook/OPT model by using method in this doc https://pytorch.org/docs/main/sparse.html#accelerating-nn-linear-with-semi-structured-sparsity

I check that the column are devisible by 8 but I got the error that:
> `/docker/pytorch/torch/sparse/semi_structured.py", line 265, in __torch_dispatch__
    result = torch._sparse_semi_structured_linear(
RuntimeError: two_four_sgemm_cutlass: Number of columns of dense matrix must be divisible by 8`

Any idea about how to solve this? Thanks

**Comment by jcaip** [comment]
@pkulium Could you please create a [forum topic](https://discuss.pytorch.org/c/uncategorized/1) or a new issue for support? Feel free to tag me and I'll take a look

**Comment by jcaip** [comment]
Also FYI, the cuSPARSELt changes have landed in main. 

You should be able to enable cuSPASRELt support by compiling from source and setting the env vars:
```
USE_CUSPARSELT=1
CUSPARSELT_ROOT=/path/to/cusparselt
```

**Comment by AboveParadise** [comment]
> > @kiucho We've added support via CUTLASS, which you should be able to use now. See these [docs](https://pytorch.org/docs/main/sparse.html#sparse-semi-structured-tensors) for more information.
> > The cuSPARSELt integration will be completed by the 2.1 release, or really when #102133 is merged.
> 
> Thanks for this work! I am trying to speed up facebook/OPT model by using method in this doc https://pytorch.org/docs/main/sparse.html#accelerating-nn-linear-with-semi-structured-sparsity
> 
> I check that the column are devisible by 8 but I got the error that:
> 
> > `/docker/pytorch/torch/sparse/semi_structured.py", line 265, in __torch_dispatch__     result = torch._sparse_semi_structured_linear( RuntimeError: two_four_sgemm_cutlass: Number of columns of dense matrix must be divisible by 8`
> 
> Any idea about how to solve this? Thanks

Same bug here, it seems that inputs.shape[0] is also supposed to be divisible by 8, which does not make sense. In Pytorch 2.1's example code, I changed input x from (3072,10240) to (3073, 10240) and got the bug.

**Comment by jcaip** [comment]
@AboveParadise @pkulium Yes there are shape constraints on both the sparse and dense matrices for both CUTLASS and CUSPARSELT. You can see [here](https://github.com/pytorch/pytorch/blob/92998693a9455af6259cae468265f01cfff8810e/torch/sparse/semi_structured.py#L16) what they are. The good news is we added padding support in #110744 so you shouldn't see an error. (But you still pay the performance cost of doing the padded matmul)

What version of PyTorch are you running?  Can you try installing the nightlies and running your code?

**Comment by RanchiZhao** [comment]
Any newest doc？ I want to use pytorch to accelerate 2:4 sparse model's inference
